mantidproject/mantid
# Prevent deadlock between observe and replace methods in ADS Observers

Fixes a deadlock that occurred when one of the observe methods of the ADS observer are called from python while a handle was being executed. This is fixed by releasing the GIL in the boost python layer before we call an observe method.

**To test:**
Two ways. Either the method described in [#30523](https://github.com/mantidproject/mantid/issues/30523) or:
1. Add a `time.sleep(2)` line to [`FigureManagerADSObserver.replaceHandle`](). This will cause the handle to sleep for two seconds.
2. Run the following script
```
from mantid.simpleapi import *

ws = CreateSampleWorkspace()
while True:
    ws = Scale(ws, 1.001, "Multiply")
```
3. Plot a spectrum from the workspace
4. Close the plot while the status bar indicates `Multiplying factor... 50%`
5. Workbench should not hang.

Fixes #30523

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
